datalad_next/runners/run.py

```diff
@@ -0,0 +1,258 @@
+            self.terminate_time = kwargs.pop("terminate_time")
+            kill_time = kwargs.pop("kill_time")
+            self.kill_time = (
+                ((self.terminate_time or 0) + kill_time)
+                if kill_time is not None
+                else kill_time
+            )
+
+            self.process: subprocess.Popen | None = None
+            self.return_code: int | None = None
+            self.kill_counter: int = 0
+
+            super().__init__(*args, **kwargs)
+
+        def arm(self) -> None:
+            self.kill_counter = 0
+            self.armed = True
+
+        def connection_made(self, process: subprocess.Popen) -> None:
+            self.process = process
+            super().connection_made(process)
+
+        def timeout(self, fd: int | None) -> bool:
+            if self.armed and fd is None:
+                self.kill_counter += 1
+                if self.kill_time and self.kill_counter >= self.kill_time:
+                    self.process.kill()
+                    self.kill_time = None
+                if self.terminate_time and self.kill_counter >= self.terminate_time:
+                    self.process.terminate()
+                    self.terminate_time = None
+
+            # If we set the timeout argument due to a not-None kill_time
+            # or a not-None terminate_time, and due to a None timeout parameter,
+            # we leave the timeout handler here.
+            if self.introduced_timeout:
+                return False
+
+            # If the timeout was set by the user of the context, we execute
+            # the timeout handler of the superclass.
+            return super().timeout(fd)
+
+        def process_exited(self) -> None:
+            self.return_code = self.process.poll()
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: type[Protocol],
+    *,
+    cwd: Path | None = None,
+    stdin: int | IO | bytes | Queue[bytes | None] | None = None,
+    timeout: float | None = None,
+    closing_action: Callable | None = None,
+    terminate_time: int | None = None,
+    kill_time: int | None = None,
+    **protocol_kwargs,
+) -> Any | Generator:
+    """ A context manager for subprocesses
+
+    The run-context manager will start a subprocess via ``ThreadedRunner``,
+    provide the result of the subprocess invocation, i.e. the result of
+    ``ThreadedRunner.run()`` in the ``as``-variable, and
+    clean up the subprocess when the context is left.
+
+    The run-context manager supports the guaranteed exit of a subprocess through
+    either:
+
+     a) natural exit of the subprocess
+     b) termination of the subprocess via SIGTERM, if ``terminate_time`` is
+        specified
+     c) termination of the subprocess via SIGKILL, if ``kill_time`` is specified
+
+    If the process terminates the run-context manager will ensure that its exit
+    status is read, in order to prevent zombie processes.
```

Comment:
This would be a good place to provide information on how the exit code could be obtained when using `run()`